### PR TITLE
Produce only a single compilation event for partial methods.

### LINKED ISSUE #7477: Duplicate compilation events for partial methods
Partial methods in C# currently result in duplicate compilation events.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #7477 

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -2885,6 +2885,10 @@ internal override AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAn
 
         internal void SymbolDeclaredEvent(Symbol symbol)
         {
+            if (symbol.Name == "PartialMethod")
+            {
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Is this test code that should be removed? Same question for VB.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs [hunk 2]
```diff
@@ -216,6 +217,7 @@ partial class Class
             Assert.True(declaredSymbolNames.Contains("N1"));
             Assert.True(declaredSymbolNames.Contains("Class"));
             Assert.True(declaredSymbolNames.Contains("NonPartialMethod1"));
+            Assert.True(declaredSymbolNames.Contains("ImpartialMethod"));
             Assert.True(declaredSymbolNames.Contains("PartialMethod"));
             Assert.True(completedCompilationUnits.Contains(tree1.FilePath));
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs]: Are we verifying the method symbol is the expected part in each case?

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 3]
```diff
@@ -626,11 +626,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                         If method.IsPartial() Then
                             Dim impl = method.PartialImplementationPart
                             If impl IsNot method Then
-                                If CType(method, SourceMethodSymbol).SetDiagnostics(ImmutableArray(Of Diagnostic).Empty) Then
-                                    method.DeclaringCompilation.SymbolDeclaredEvent(method)
-                                End If
+                                CType(method, SourceMethodSymbol).SetDiagnostics(ImmutableArray(Of Diagnostic).Empty)
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: It's not clear to me why the `If` around `SetDiagnostics()` is unnecessary. Consider adding a comment.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 4]
```diff
@@ -1193,7 +1194,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                                 End If
                                 Return semanticModel
                             End Function)
-                        compilation.EventQueue.Enqueue(New SymbolDeclaredCompilationEvent(compilation, method, lazySemanticModel))
+                        If method.Name = "PartialMethod" Then
+                            method = method
+                        End If
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Test code?

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 5]
```diff
@@ -948,7 +948,8 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                         return semanticModel;
                     });
 
-                    _compilation.EventQueue.Enqueue(new SymbolDeclaredCompilationEvent(_compilation, methodSymbol, lazySemanticModel));
+                    MethodSymbol symbolToProduce = methodSymbol.PartialDefinitionPart ?? methodSymbol;
+                    _compilation.EventQueue.Enqueue(new SymbolDeclaredCompilationEvent(_compilation, symbolToProduce, lazySemanticModel));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: I am not sure this code is doing the right thing now. The partial definition can be in a different syntax tree, but we are passing semantic model for implementation part.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 6]
```diff
@@ -2381,6 +2381,11 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
                     {
                         diagnostics.Add(ErrorCode.ERR_PartialMethodMustHaveLatent, method.Locations[0], method);
                     }
+
+                    if (method.IsPartialDefinition && (object)method.PartialImplementationPart == null)
+                    {
+                        method.DeclaringCompilation.SymbolDeclaredEvent(method);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Is it safe to raise this event at this point. It looks like we might be in a state where we haven't finished building member list for the containing type.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: What is the purpose of calling SetDiagnostics with an empty array? Does it do something useful?

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 7]
```diff
@@ -1180,7 +1179,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Dim compilation = compilationState.Compilation
                 If compilation.ShouldAddEvent(method) Then
                     If block Is Nothing Then
-                        compilation.SymbolDeclaredEvent(sourceMethod)
+                        If Not method.IsPartial OrElse method.PartialImplementationPart Is Nothing Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Can we even get here with the partial definition method? I think we always switch to implementation before calling CompileMethod. If this is correct, I would prefer that we add an assert to that fact at the beginning of the method and revert this change.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 8]
```diff
@@ -1193,7 +1194,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                                 End If
                                 Return semanticModel
                             End Function)
-                        compilation.EventQueue.Enqueue(New SymbolDeclaredCompilationEvent(compilation, method, lazySemanticModel))
+                        If method.Name = "PartialMethod" Then
+                            method = method
+                        End If
+                        Dim symbolToProduce = If(method.PartialDefinitionPart, method)
+                        compilation.EventQueue.Enqueue(New SymbolDeclaredCompilationEvent(compilation, symbolToProduce, lazySemanticModel))
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: The same issue with possible inconsistency between the symbol and semantic model.

**Comment by AlekseyTs** [comment]
In order to do a proper review of this change, I think we need to put in words precisely what events are expected to be raised and with what data for partial methods in each of the following scenarios:
- Both declaration and implementation is present
- Only declaration is present
- Only implementation is present

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb (hunk 7 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Yes, we get here if there is no implementation part.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb (hunk 3 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Yes, SetDiagnostics marks the diagnostics for the body as being set.

Moving the event back into the if.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: You have to give the compilation event a syntax reference in order to get back a semantic model, so it will only give you this one if you're asking about the implementation part. Since the declaration has *locations* for both parts, this is the one we need to cache to get the best performance.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Yes, but that's OK. If a receiver of the event attempts to do anything with the member list, they will have to wait for it to be completed and stored. Since they are necessarily on a different thread, no problem.

**Comment by gafter** [comment]
@AlekseyTs In either of the first two cases, only the declaration part. The third case cannot occur for VB. The third case should result in a symbol event for C#. We have tests for all three cases.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Would it be possible to simply adjust original condition in LazyAsyncMethodChecks instead?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I don't understand what you are suggesting. Which condition, how are you suggesting it be adjusted, and instead of what?

Events for partial methods have to be delayed so that the symbol is not exposed until the partial definition and the partial declaration have been paired up.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review. I'm not aware of any outstanding actionable issues with this PR, and it is targeted for 1.2.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: What I mean is I would be more comfortable if instead of moving code from src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs (line 395) to src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (line 2385) the change simply adjusted condition at the original place.

Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Done.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review. I'm not aware of any outstanding actionable issues with this PR, and it is targeted for 1.2.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: Could you please add an assert at the beginning of the method that it is never called for a partial definition?

**Comment by AlekseyTs** [comment]
Could you please add an assert at the beginning of CompileMethod in (...rs/VisualBasic/Portable/Compilation/MethodCompiler.vb) that it is never called for a partial definition? While responding to my feedback you reverted changes to the method, but didn't add the assert.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
LGTM
